### PR TITLE
KAFKA-15281: Implement the groupMetadata Consumer API

### PR DESCRIPTION
**Pull request by cadonna**
Implements the groupMetadata() API on the async consumer.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 1]
```diff
@@ -131,9 +135,78 @@ public class AsyncKafkaConsumer<K, V> implements ConsumerDelegate<K, V> {
 
     private static final long NO_CURRENT_THREAD = -1L;
 
+    /**
+     * An {@link org.apache.kafka.clients.consumer.internals.events.EventProcessor} that is created and executes in the application thread for the purpose of processing
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Super nit: reflow the comments around whatever the line length standard is.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 2]
```diff
@@ -398,6 +466,38 @@ public class AsyncKafkaConsumer<K, V> implements ConsumerDelegate<K, V> {
...
+            config.ignore(THROW_ON_FETCH_STABLE_OFFSET_UNSUPPORTED);
+        }
+        return groupMetadata;
+    }
+
+    private Optional<ConsumerGroupMetadata> initializeGroupMetadata(final String groupId,
+                                                                    final Optional<String> groupInstanceId) {
+        if (groupId != null) {
+            if (groupId.isEmpty()) {
+                throwInInvalidGroupIdException();
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Why are we throwing this error here?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 3]
```diff
@@ -849,9 +953,22 @@ public OptionalLong currentLag(TopicPartition topicPartition) {
         }
     }
 
+    public void setGroupMetadata(final ConsumerGroupMetadata groupMetadata) {
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: This is just for testing, right? Can we remove the `public` qualifier?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 4]
```diff
@@ -849,9 +953,22 @@ public OptionalLong currentLag(TopicPartition topicPartition) {
...
+        this.groupMetadata = Optional.of(groupMetadata);
+    }
+
     @Override
     public ConsumerGroupMetadata groupMetadata() {
-        throw new KafkaException("method not implemented");
+        acquireAndEnsureOpen();
+        try {
+            maybeThrowInvalidGroupIdException();
+            backgroundEventProcessor.process();
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: I don't think we want to process _all_ the events here, as this might process events (like async commits, et al.) that are unrelated to group metadata.

I have this same issue in another PR I'm working on, so I don't have a solution yet 😞

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 5]
```diff
@@ -849,9 +953,22 @@ public OptionalLong currentLag(TopicPartition topicPartition) {
...
+            return groupMetadata.orElseThrow(
+                () -> new IllegalStateException("No group metadata found although a valid group ID exists. This is a bug!")
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: I'm not sure what we should throw here. Does the API contract allow for `IllegalStateException`?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 6]
```diff
@@ -1365,7 +1482,11 @@ public KafkaConsumerMetrics kafkaConsumerMetrics() {
     private void maybeThrowFencedInstanceException() {
         if (isFenced) {
             throw new FencedInstanceIdException("Get fenced exception for group.instance.id " +
-                groupInstanceId.orElse("null"));
+                groupMetadata.orElseThrow(
+                    () -> new IllegalStateException("No group metadata found although a group ID was provided. This is a bug!")
+                ).groupInstanceId().orElseThrow(
+                    () -> new IllegalStateException("No group instance ID found although the consumer is fenced. This is a bug!")
+                ));
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Can we restructure this so that our exception throwing case doesn't have nested exception throwing cases?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/BackgroundEventHandler.java [hunk 7]
```diff
@@ -49,4 +49,9 @@ public void add(BackgroundEvent event) {
         log.trace("Enqueued event: {}", event);
         backgroundEventQueue.add(event);
     }
+
+    // Visible for testing
+    public Queue<BackgroundEvent> backgroundEventQueue() {
+        return backgroundEventQueue;
+    }
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/BackgroundEventHandler.java]: I'm not terribly crazy about this approach as it can lead to circumvention of the logic. I understand it's just for tests, for which we do have the `ConsumerTestBuilder` approach which allows tests to get access to these kinds of objects directly.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/GroupMetadataUpdatedEvent.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) return true;
+    if (o == null || getClass() != o.getClass()) return false;
+    if (!super.equals(o)) return false;
+
+    GroupMetadataUpdatedEvent that = (GroupMetadataUpdatedEvent) o;
+
+    return groupMetadata.equals(that.groupMetadata);
```
Comment (kirktrue) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/GroupMetadataUpdatedEvent.java]: I just double-checked that `ConsumerGroupMetadata` has valid `hashCode()` and `equals()` methods defined, and it does, so we're 👍

**Review by kirktrue — CHANGES_REQUESTED** [review]
Thanks for the PR, @cadonna!

It looks like there is some work left on the tests. The intention of the `ConsumerTestBuilder` was to be able to construct a working `AsyncKafkaConsumer` with access to most/all of the objects used therein. It appears, however, that there's some complexity with the background event queue that is making your life harder than it should. Can you elaborate on the direction to remove the background queue from the 'test builder' instead of using the one it constructed?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 2 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Because the group ID cannot be empty if it is set.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 3 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: That was a temporary change that I missed to remove.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 9]
```diff
@@ -666,12 +766,16 @@ public Map<TopicPartition, OffsetAndMetadata> committed(final Set<TopicPartition
...
     private void maybeThrowInvalidGroupIdException() {
-        if (!groupId.isPresent() || groupId.get().isEmpty()) {
-            throw new InvalidGroupIdException("To use the group management or offset commit APIs, you must " +
-                    "provide a valid " + ConsumerConfig.GROUP_ID_CONFIG + " in the consumer configuration.");
+        if (!groupMetadata.isPresent()) {
+            throwInInvalidGroupIdException();
         }
     }
 
+    private void throwInInvalidGroupIdException() {
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: I find the name `throwInInvalidGroupIdException` a bit strange. `throwInvalidGroupIdException` seems more usual.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 4 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: OK, should then `backgroundEventProcessor.process()` only be called in `AsyncKafkaConsumer#poll()`?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/BackgroundEventHandler.java (hunk 7 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/BackgroundEventHandler.java]: I had to take painkillers to stand the pain when I added that method because I totally agree with you but I did not find another way. I will reconsider.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 5 above)
Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: You *know* that groupMetadata is present because of the earlier `maybeThrowInvalidGroupIdException`. I suppose one pattern would be to return an unwrapped `GroupMetadata` from maybeThrowInvalidGroupIdException so that you've eliminated the possibility of the `Optional` being empty.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 10]
```diff
@@ -1365,7 +1482,11 @@ public KafkaConsumerMetrics kafkaConsumerMetrics() {
     private void maybeThrowFencedInstanceException() {
         if (isFenced) {
             throw new FencedInstanceIdException("Get fenced exception for group.instance.id " +
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Please use the `ConsumerConfigs` constant for the `"group.instance.id"`.

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/BackgroundEventHandlerTest.java [hunk 11]
```diff
@@ -51,91 +61,91 @@ public void tearDown() {
...
-            backgroundEventProcessor.process();
-            fail("Should have thrown error: " + error);
-        } catch (Throwable t) {
-            assertEquals(error.getClass(), t.getClass());
-            assertEquals(error.getMessage(), t.getMessage());
-        }
-
-        assertTrue(backgroundEventQueue.isEmpty());
-    }
+//    @Test
```
Comment (AndrewJSchofield) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/BackgroundEventHandlerTest.java]: I suppose the final PR will not have all of these commented lines.

**Review by AndrewJSchofield — COMMENTED** [review]
lgtm with some minor comments. I like the restructuring to enclose the background event processor.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 9 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Yeah, that was already fixed in https://github.com/apache/kafka/pull/14872

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 10 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Touché 🙂

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/BackgroundEventHandlerTest.java (hunk 11 above)
Comment (cadonna) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/events/BackgroundEventHandlerTest.java]: You need to have those commented out code in a draft PR, otherwise GitHub let's you not mark it as draft 🙂 

Jokes aside, yes they will be removed.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 9 above)
Comment (AndrewJSchofield) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: What a tangled web we weave.

**Comment by cadonna** [comment]
> Can you elaborate on the direction to remove the background queue from the 'test builder' instead of using the one it constructed?

I had issues with tests using the spy on the `AsyncKafkaConsumer`. More precisely, a test failed with the spy but did not fail with a consumer not wrapped into a spy. 
BTW, IMO, spies (or partial mocks) should be used really carefully. Actually, good code does not need spies (with a few exceptions). Spies avoid to structure the code well. They do not force one to loosely couple objects. Even the [Mockito documentation](https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#spy) warns against their own spies. Additionally, the code under test, i.e., the async Kafka consumer, should not be wrapped into a spy. We should test that code directly to avoid possible side effects coming from the wrapping or from mistakes in specifying stubs on the spy.

**Comment by AndrewJSchofield** [comment]
> > Can you elaborate on the direction to remove the background queue from the 'test builder' instead of using the one it constructed?
> 
> I had issues with tests using the spy on the `AsyncKafkaConsumer`. More precisely, a test failed with the spy but did not fail with a consumer not wrapped into a spy. BTW, IMO, spies (or partial mocks) should be used really carefully. Actually, good code does not need spies (with a few exceptions). Spies avoid to structure the code well. They do not force one to loosely couple objects. Even the [Mockito documentation](https://javadoc.io/doc/org.mockito/mockito-core/latest/org/mockito/Mockito.html#spy) warns against their own spies. Additionally, the code under test, i.e., the async Kafka consumer, should not be wrapped into a spy. We should test that code directly to avoid possible side effects coming from the wrapping or from mistakes in specifying stubs on the spy.

I've had exactly the same problem. Nested mocking makes it all very unhappy.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 5 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Actually, `backgroundEventProcessor.process();` sets the group metadata. So returning group metadata from `maybeThrowInvalidGroupIdException` might return an outdated group metadata. Since now I removed `backgroundEventProcessor.process();`. I will reconsider the exception.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 4 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: I think we should leave it only in `poll()` for the time being, yes 😞

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 2 above)
Comment (kirktrue) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Oh, the line after `throwInInvalidGroupIdException` is just to appease the compiler?

Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Yes, but I changed also the code to get rid of that line.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/BackgroundEventHandler.java (hunk 7 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/BackgroundEventHandler.java]: OK, was able to get rid of that abominable piece of code.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/GroupMetadataUpdateEvent.java [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+    final private String groupId;
+    final private int memberEpoch;
+    final private String memberId;
+    final private Optional<String> groupInstanceId;
+
+    public GroupMetadataUpdateEvent(final String groupId,
+                                    final int memberEpoch,
+                                    final String memberId,
+                                    final Optional<String> groupInstanceId) {
+        super(Type.GROUP_METADATA_UPDATED);
```
Comment (AndrewJSchofield) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/GroupMetadataUpdateEvent.java]: Sorry for being a naming fundamentalist, but you've used `GroupMetadataUpdateEvent` but `GROUP_METADATA_UPDATED`. Please use the same tense for the class and the constant. I'm sorting out the subclasses of `ApplicationEvent` in a similar vein in another PR. Just makes it slightly easier to navigate the code.

**Review by AndrewJSchofield — APPROVED** [review]
Approved with one tiny nit.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/GroupMetadataUpdateEvent.java (hunk 12 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/events/GroupMetadataUpdateEvent.java]: Totally agree! I renamed `GroupMetadataUpdateEvent` from `GroupMetadataUpdatedEvent` and missed to rename `GROUP_METADATA_UPDATED`.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 13]
```diff
@@ -131,6 +133,77 @@ public class AsyncKafkaConsumer<K, V> implements ConsumerDelegate<K, V> {
...
+
+        private void process(final ErrorBackgroundEvent event) {
+            throw event.error();
+        }
+
+        private void process(final GroupMetadataUpdateEvent event) {
+            if (AsyncKafkaConsumer.this.groupMetadata.isPresent()) {
+                final ConsumerGroupMetadata currentGroupMetadata = AsyncKafkaConsumer.this.groupMetadata.get();
+                AsyncKafkaConsumer.this.groupMetadata = Optional.of(new ConsumerGroupMetadata(
+                    event.groupId(),
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: So `groupId` in `this.groupMetadata` comes from `GroupRebalanceConfig.groupId` the `config.getString(CommonClientConfigs.GROUP_ID_CONFIG)`.

and `groupId` in `event.groupId` comes from `MembershipManagerImpl.groupId` which comes from `GroupState.groupId` which comes from `GroupRebalanceConfig.groupId`.

So, aren't you always overwriting the group Id with itself here?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 14]
```diff
@@ -131,6 +133,77 @@ public class AsyncKafkaConsumer<K, V> implements ConsumerDelegate<K, V> {
...
+                    event.memberEpoch(),
+                    event.memberId() != null ? event.memberId() : currentGroupMetadata.memberId(),
+                    event.groupInstanceId()
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Same as above, we read it from the config, does it ever change?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java [hunk 15]
```diff
@@ -131,6 +133,77 @@ public class AsyncKafkaConsumer<K, V> implements ConsumerDelegate<K, V> {
...
+            throw event.error();
+        }
+
+        private void process(final GroupMetadataUpdateEvent event) {
+            if (AsyncKafkaConsumer.this.groupMetadata.isPresent()) {
+                final ConsumerGroupMetadata currentGroupMetadata = AsyncKafkaConsumer.this.groupMetadata.get();
+                AsyncKafkaConsumer.this.groupMetadata = Optional.of(new ConsumerGroupMetadata(
+                    event.groupId(),
+                    event.memberEpoch(),
+                    event.memberId() != null ? event.memberId() : currentGroupMetadata.memberId(),
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: I'd slightly prefer this logic to be resolved on the background thread and make CurrentGroupMetadataEvent.memberId always non-null

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java [hunk 16]
```diff
@@ -232,6 +233,12 @@ private void onResponse(final ConsumerGroupHeartbeatResponse response, long curr
             this.heartbeatRequestState.onSuccessfulAttempt(currentTimeMs);
             this.heartbeatRequestState.resetTimer();
             this.membershipManager.onHeartbeatResponseReceived(response.data());
+            this.backgroundEventHandler.add(new GroupMetadataUpdateEvent(
```
Comment (lucasbru) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: Could we only send this if the member epoch or member ID changes? I think it's somewhat inelegant that we enqueue this for every single heartbeat. For example, we had the default max.poll.interval.ms = 300 seconds between all polls and the default heartbeat.interval= 3 seconds for each heartbeat, means we have 100 metadata update events to go through in each poll.

**Review by lucasbru — COMMENTED** [review]
thanks for the PR! I left some comments

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 13 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: Yes, we do. We could verify that the group ID and the group instance ID are equal to catch bugs. Or we could just send the updateable fields to the application thread.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java (hunk 14 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/AsyncKafkaConsumer.java]: See above

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java (hunk 16 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: Yeah, I wanted to leave that as improvement, but I think it is simple enough to include it immediately.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java [hunk 17]
```diff
@@ -232,6 +235,16 @@ private void onResponse(final ConsumerGroupHeartbeatResponse response, long curr
...
+            if (previousGroupMetadataUdateEvent == null ||
+                previousGroupMetadataUdateEvent.memberEpoch() != membershipManager.memberEpoch()) {
+
+                final GroupMetadataUpdateEvent currentGroupMetadataUpdateEvent = new GroupMetadataUpdateEvent(
+                    membershipManager.memberEpoch(),
+                    previousGroupMetadataUdateEvent == null ? membershipManager.memberId() : previousGroupMetadataUdateEvent.memberId()
+                );
+                this.backgroundEventHandler.add(currentGroupMetadataUpdateEvent);
+                previousGroupMetadataUdateEvent = currentGroupMetadataUpdateEvent;
+            }
```
Comment (cadonna) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: @dajac @lucasbru @kirktrue @AndrewJSchofield Could you please sanity check this code snippet?
The heartbeat request manager should only send an group metadata update event if it is the first heartbeat response it gets (i.e. `previousGroupMetadataUdateEvent == null`), because that sets the member ID and the member ID never changes after this. Or the member epoch changed in the last heartbeat response (i.e., previousGroupMetadataUdateEvent.memberEpoch() != membershipManager.memberEpoch()).

Does this make sense?

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java [hunk 18]
```diff
@@ -292,6 +297,90 @@ public void testValidateConsumerGroupHeartbeatRequest(final short version) {
...
+        NetworkClientDelegate.PollResult result = heartbeatRequestManager.poll(time.milliseconds());
+        assertEquals(1, result.unsentRequests.size());
+        NetworkClientDelegate.UnsentRequest request = result.unsentRequests.get(0);
+        ClientResponse firstResponse = createHeartbeatResponse(request, Errors.NONE);
+        request.handler().onComplete(firstResponse);
+        assertEquals(1, backgroundEventQueue.size());
+        final BackgroundEvent firstEvent = backgroundEventQueue.poll();
+        assertEquals(BackgroundEvent.Type.GROUP_METADATA_UPDATE, firstEvent.type());
+
+        time.sleep(2000);
```
Comment (cadonna) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:315]: @kirktrue Although I call  `resetWithZeroHeartbeatInterval()` in the beginning of this test method, I need to sleep at least 1000 ms to get a second heartbeat response. I thought `resetWithZeroHeartbeatInterval()` sets the heartbeat interval to zero, but during debugging I learnt that the heartbeat interval is still 1000 ms. Is this intended?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java (hunk 17 above)
Comment (lucasbru) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: Can't the `memberId` change when I rejoin the group?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java (hunk 16 above)
Comment (lucasbru) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: creating a little follow-up ticket would have been fine as well, but looking good.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java (hunk 17 above)
Comment (cadonna) [reply clients/src/main/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManager.java]: I think you are right! Let me change that.

**Review by lucasbru — APPROVED** [review]
LGTM, thanks!

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java (hunk 18 above)
Comment (kirktrue) [reply clients/src/test/java/org/apache/kafka/clients/consumer/internals/HeartbeatRequestManagerTest.java:315]: That's the intention of `resetWithZeroHeartbeatInterval()`, yes. At some point it did as it said 🤔
